pkg/util/admission/work_queue.go

```diff
@@ -1531,11 +1549,61 @@ type WorkQueueMetrics struct {
 	WaitQueueLength *metric.Gauge
 }
 
+func (m *WorkQueueMetrics) incRequested(priority admissionpb.WorkPriority) {
+	m.total.Requested.Inc(1)
+	m.getOrCreate(priority).Requested.Inc(1)
+}
+
+func (m *WorkQueueMetrics) incAdmitted(priority admissionpb.WorkPriority) {
+	m.total.Admitted.Inc(1)
+	m.getOrCreate(priority).Admitted.Inc(1)
+}
+
+func (m *WorkQueueMetrics) incErrored(priority admissionpb.WorkPriority) {
+	m.total.Errored.Inc(1)
+	m.getOrCreate(priority).Errored.Inc(1)
+}
+
+func (m *WorkQueueMetrics) recordStartWait(priority admissionpb.WorkPriority) {
+	m.total.WaitQueueLength.Inc(1)
+	m.getOrCreate(priority).WaitQueueLength.Inc(1)
+}
+
+func (m *WorkQueueMetrics) recordFinishWait(priority admissionpb.WorkPriority, dur time.Duration) {
+	m.total.WaitQueueLength.Dec(1)
+	m.total.WaitDurations.RecordValue(dur.Nanoseconds())
+	m.total.WaitDurationSum.Inc(dur.Microseconds())
+
+	priorityStats := m.getOrCreate(priority)
+	priorityStats.WaitQueueLength.Dec(1)
+	priorityStats.WaitDurations.RecordValue(dur.Nanoseconds())
+	priorityStats.WaitDurationSum.Inc(dur.Microseconds())
+}
+
 // MetricStruct implements the metric.Struct interface.
-func (WorkQueueMetrics) MetricStruct() {}
+func (*WorkQueueMetrics) MetricStruct() {}
+
+func makeWorkQueueMetrics(
+	name string, registry *metric.Registry, applicablePriorities ...admissionpb.WorkPriority,
+) *WorkQueueMetrics {
+	totalMetric := makeWorkQueueMetricsSingle(name)
+	registry.AddMetricStruct(totalMetric)
+	wqm := &WorkQueueMetrics{
+		name:     name,
+		total:    totalMetric,
+		registry: registry,
+	}
+	// TODO(abaptist): This is done to pre-register stats. Need to check that we
+	// getOrCreate "enough" of the priorities to be useful.
+	for _, pri := range applicablePriorities {
```

Comment:
So it seems we're doing pre-registration here anyway. Do we need the LoadOrStore above when finding the right metric? If everything is properly pre-registered, shouldn't we fatal if we don't find something in the sync.Map?